tests/integration/test_l1_read_request.py

```diff
@@ -0,0 +1,283 @@
+import re
+import unittest
+from callee import Regex
+from common.constants import MESIState, LogLevel
+from utils.event_handler import handle_event
+from tests.integration.integration_setup import IntegrationSetup
+
+
+class TestCommandL1ReadRequestData(IntegrationSetup):
+
+    def setUp(self):
+        super().setUp()
+        # Default to 0 for trace event op code.
+        # This allows us to create a child class for the instruction read requests,
+        # which should have identical behavior.
+        self.trace_event = 0
+        # nohit_addr is an address that results in a simulated
+        # NOHIT snoop response from other caches
+        self.nohit_addr = self.nohit_addresses[0]
+        # hit_addr is an address that results in a simulated
+        # HIT snoop response from other caches
+        self.hit_addr = self.hit_addresses[0]
+        # hitm_addr is an address that results in a simulated
+        # HITM snoop response from other caches
+        self.hitm_addr = self.hitm_addresses[0]
+
+    def test_exclusive(self):
+        """
+        Test reading a line in the Exclusive state.
+        Other caches return a NOHIT snoop response.
+
+        Requires the following sequence:
+        - Read line (cache miss)
+        - Read line (cache hit)
+        """
+        # Increment expected stats values based on actions we expect to
+        # modify the cache statistics object
+        # Note: these values are not determined by the actual return value
+        # of the functions being called
+        expected_reads = 0
+        expected_writes = 0
+        expected_hits = 0
+        expected_misses = 0
+
+        # Event 0: Read miss at self.nohit_addr (gets line in E state)
+        handle_event(self.cache, self.trace_event, self.nohit_addr)
+        # FIXME: Enable this one we've updated the code to send this message
+        # self.mock_logger.assert_called_with(
+        #     LogLevel.NORMAL, Regex(r".*L2.*sendline.*")
+        # )
```

Comment:
Just FYI, this is an example of a section we need to re-enable after making changes in main source code. We might change the exact call to use the `assert_log_called_once_with` helper, but this is mainly here so we know we need to come back to this.

Reply:
@nkanderson I think you already know about my fixes for this in my current PR #52 but I've added the L1 messaging logic so once that is merged we should be good to go.

Reply:
Ah right, yeah, since I was working off of `main`, I guess I was focused on what was failing with the current code. So should be a quick change to these sections once both your PR and this one are merged.

I think we still need to add the sendline for write events, is that right? I can't remember if you added that in that other PR or not.

Reply:
I added the sendline for write events but I'm still a little confused on when we should message L1 since it's a write through on first write. I'm not sure how we can tell if a write request is a miss or hit in L1. Maybe its a good assumption to send the line every-time no matter what since we don't know what is in L1. Either way this has been added but we might need to revisit because there is one case I might have forgot to included, depending on what our final assumption is.

Reply:
> I'm still a little confused on when we should message L1 since it's a write through on first write. I'm not sure how we can tell if a write request is a miss or hit in L1

I posted on canvas asking about that, if I'm understanding your description correctly. We won't know if the write from L1 is due to a local eviction in L1 or a write miss in L1, so we send the line for any L1 write request. Since it's write-through on first write, we just need to make sure L2 has the new data from any L1 write - this aspect is a little confusing, since we aren't actually dealing with any data in our simulation, but for our purposes I think just means we need to make sure L2 has allocated the line on any L1 write.